File: 0322 - Coin Change/solution.py
from functools import lru_cache
class Solution:
    def coinChange(self, A: List[int], n: int) -> int:
        @lru_cache(None)
        def dp(k):
            if k == 0:
                return 0
            if all(k < i for i in A):
                return n+1
            return 1+min(dp(k-i) for i in A if k >= i)
        return -1 if dp(n) >= n+1 else dp(n)

File: 0322 - Coin Change/solution.cpp
class Solution {
public:
    int coinChange(vector<int>& A, int n) {
        vector<int> dp(n+1, n+1); dp[0] = 0;
        for(int i = 1; i <= n; i++)
            for(auto j : A)
                if(i >= j) dp[i] = min(dp[i], 1+dp[i-j]);
        return dp[n]==n+1?-1:dp[n];
    }
};